### DISCUSSION TITLE
Unexpected error launching SEB with full dependencies installed.

### DISCUSSION THREAD

**Discussion by acardielf** [SafeExamBrowser/seb-win-refactoring#1003] [Q&A] [answered]
We're having some problems in certain PC running SEB over Windows 11. We used the SetupBundle installator which include runtimes and other kind of dependencies. But we're getting this error:

```
2024-10-17 12:53:54.478 [01] - ERROR: Caught unexpected exception while performing operation 'LazyInitializationOperation'!

   Exception Message: No se puede cargar el archivo o ensamblado 'CefSharp.Core.Runtime.dll' ni una de sus dependencias. Acceso denegado.
   Exception Type: System.IO.FileLoadException

   en CefSharp.CefSettingsBase..ctor()
   en SafeExamBrowser.Browser.BrowserApplication.InitializeCefSettings() en C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:línea 332
   en SafeExamBrowser.Browser.BrowserApplication.Initialize() en C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:línea 112
   en SafeExamBrowser.Client.Operations.BrowserOperation.Perform() en C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Client\Operations\BrowserOperation.cs:línea 47
   en SafeExamBrowser.Core.OperationModel.OperationSequence.Perform() en C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Core\OperationModel\OperationSequence.cs:línea 108
```

I'm pretty sure the system has the .NET libraries dependencies installed (as a I say, we used the SetupBundle version and I also re-install manually the required dependencies). 

The weird thing is the next: The Windows user session who is running SEB has Administration permissions in the system. If I execute SEB, forcing "Run As Administrator", the software is launching properly, but, if I give to the students a "config.seb" file, and them make double-click, SEB is not running as Administrator (UAC is not showing). 

The solution for me, and for this users was create a custom seb_launcher.bat with this content: 

```
@echo off
setlocal
set SEB_PATH="%programfiles%\SafeExamBrowser\SafeExamBrowser.exe"
set SEB_FILE=%1
powershell -Command "Start-Process '%SEB_PATH%' -ArgumentList '%SEB_FILE%' -Verb RunAs"
```

And use the context menu to "Run with" the seb_launcher.bat to ensure the application is running with Administrator permissions. 

Despite of I "solved" the problem, is not so logical this way to make it work. So, can someone give me some clue about how to solve it? 

[2024-10-17_12h53m50s_Client.log](https://github.com/user-attachments/files/17489872/2024-10-17_12h53m50s_Client.log)

**Comment by dbuechel** [comment]
SEB for Windows should indeed not require administrator privileges to run correctly. The error message referring to CefSharp libraries indicates normally that the Visual C++ Redistributable is not installed or that there is a version mismatch. Thus, please verify that the correct version as specified in the release notes is installed.

**Reply by acardielf** [reply]
I'm totally sure the Visual C++ Redistributable is installed in the system; I did manually, and a I read carefully the release notes and the documentation. In fact, the application is running 100% properly if I launch it with Administrator privileges, which is not the deseable option, but if it would miss, it cannot be run properly. Maybe it's related to permissions or something like that, but the user belong to Administrators group in the system. That situation occurs in 5 of 30 PCs with a just installed version of Windows 11 three months ago.

**Reply by dbuechel** [reply]
> Maybe it's related to permissions or something like that, but the user belong to Administrators group in the system.

That might very well be and is what I suspect as well; as stated, SEB for Windows should not require administrator privileges in order to run correctly.

**Reply by acardielf** [reply]
Maybe is because there other service or application accessing to the same file at the same time?

**Reply by dbuechel** [reply]
Yes, albeit that it would need to be some other system or third-party application, as SEB itself has a single-instance mechanism which ensures that there cannot be multiple SEBs running at the same time.

**Comment by jbas007** [comment]
Hi all,

I have a student with the same issue, only that running as an Administrator does not work. Confirmed all correct dependencies are installed on the system (Windows 11). Interestingly, launch only fails when opening a config file, launching from inside Moodle, or when specifying Moodle in the start URL. When opening SEB from applications (and not specifying a start URL in the preferences) there is no issue. Any other ideas what may be causing this? 

2024-11-08 10:37:44.819 [01] - ERROR: Caught unexpected exception while performing operation 'LazyInitializationOperation'!
   Exception Message: Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.
   Exception Type: System.IO.FileNotFoundException
   at CefSharp.CefSettingsBase..ctor()
   at SafeExamBrowser.Browser.BrowserApplication.InitializeCefSettings() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:line 332
   at SafeExamBrowser.Browser.BrowserApplication.Initialize() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:line 112
   at SafeExamBrowser.Client.Operations.BrowserOperation.Perform() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Client\Operations\BrowserOperation.cs:line 47
   at SafeExamBrowser.Core.OperationModel.OperationSequence.Perform() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Core\OperationModel\OperationSequence.cs:line 108

**Reply by dbuechel** [reply]
With which SEB version are you experiencing the issue?

**Reply by jbas007** [reply]
Version 3.8.0

Note that logs show the following error after the one above:

2024-11-08 10:37:44.819 [01] - INFO: Terminating browser...
2024-11-08 10:37:44.819 [01] - INFO: [BrowserApplication] Initiating termination...
2024-11-08 10:37:45.337 [01] - ERROR: Caught unexpected exception while reverting operation 'LazyInitializationOperation'!
   Exception Message: Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.
   Exception Type: System.IO.FileNotFoundException
   **at CefSharp.Cef.GetGlobalCookieManager(ICompletionCallback callback)**
   at SafeExamBrowser.Browser.BrowserApplication.DeleteCookies() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:line 278
   at SafeExamBrowser.Browser.BrowserApplication.Terminate() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Browser\BrowserApplication.cs:line 166
   at SafeExamBrowser.Client.Operations.BrowserOperation.Revert() in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Client\Operations\BrowserOperation.cs:line 76
   at SafeExamBrowser.Core.OperationModel.OperationSequence.Revert(Boolean regress) in C:\Users\appveyor\projects\seb-win-refactoring-t7rrnbsqeloke9rp\SafeExamBrowser.Core\OperationModel\OperationSequence.cs:line 136

**Reply by dbuechel** [reply]
Yes, again, do make sure that you've correctly installed all the required runtime dependencies as specified in the release notes or simply use the setup bundle for installation.

**Reply by acardielf** [reply]
> Yes, again, do make sure that you've correctly installed all the required runtime dependencies as specified in the release notes or simply use the setup bundle for installation.

That's what I did, sorry for the insistence :(

**Reply by dbuechel** [reply]
Terribly sorry, I was referring to @jbas007.

**Reply by jbas007** [reply]
Thanks @dbuechel ! I can confirm that these students have all the required dependencies installed, and are also using the setup bundle.

**Reply by jbas007** [reply]
Hi all! An update from my end - C++ was installed, but needed repairing through the installer. This fixed the problem. Thank you both for your help!

**Reply by acardielf** [reply]
> Hi all! An update from my end - C++ was installed, but needed repairing through the installer. This fixed the problem. Thank you both for your help!

May I know what exactly update you installed to fix the problem?

**Reply by jbas007** [reply, marked as answer]
It wasn't an update as such. Visual C++ 2015-2019 Redistributable (x64) v14.26.28720 was installed already but we opened the installer through Control Panel -> Programs and Features, and selected 'Repair'. They could then use SEB without issue.

**Reply by dbuechel** [reply]
Thanks for the update @jbas007 - that is indeed something along the lines of what was to be expected.

**Comment by acardielf** [comment]
I'm still getting the same error with version 3.8.0 build 742 running over Windows 11 (Microsoft Windows NT 10.0.22621.0)

[2024-11-12_12h23m39s_Runtime.log](https://github.com/user-attachments/files/17715772/2024-11-12_12h23m39s_Runtime.log)
[2024-11-12_12h23m39s_Client.log](https://github.com/user-attachments/files/17715773/2024-11-12_12h23m39s_Client.log)

But the error appears on 6 PCs out of 30.

**Reply by dbuechel** [reply]
The only other thing that comes to my mind apart from system (policies) or security software would be another application also using CefSharp or CEF, as their usage is quite ubiquitous.

But if it's not the runtime dependencies nor another application interfering, then you might have more luck creating an issue in the upstream repository of the browser engine, see https://github.com/cefsharp/CefSharp/issues.

**Reply by acardielf** [reply]
I did some research with tools like `handle` from SysInternals or Powershell ( `Get-Process | Where-Object { $_.Modules | Where-Object { $_.FileName -eq "<path>" } }` with no success. No other process seems accessing that file. I will keep looking for a solution; if I found something I'll inform about it in this issue.

**Reply by dbuechel** [reply]
Great, thanks a lot - that would indeed be very kind and of great interest for us as well as the community!

**Reply by jbas007** [reply]
Thanks @ancafe !

**Comment by acardielf** [comment]
I have some news about this issue. After few weeks I discovered other similar error with other software (VirtualBox). As we commented in this issue, it was related to file permissions. Repair or reinstall C++ Redistributable packages didn't work. 
Finally, I repair the system file permissions executing this command with a Administrator permissions console, and it works! 

```powershell
ICACLS * /T /Q /C /RESET
```
Maybe it's a very aggressive solution, but it fix me the error on all failing computers. 

So, from my part, the issue it's solved!

Thanks for your support, and my best wishes for the new year.

**Reply by dbuechel** [reply]
Great, that is excellent to know, thanks for sharing the information. How the file permissions got corrupted would be good to know, in case you happen to know that or have a suspicion.
